crates/sui-core/src/authority/authority_per_epoch_store_util.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::authority_per_epoch_store::DeferralKey;
+use core::cmp::max;
+use narwhal_types::Round;
+use std::collections::HashMap;
+use sui_types::base_types::{ObjectID, TransactionDigest};
+use sui_types::executable_transaction::VerifiedExecutableTransaction;
+use sui_types::transaction::SharedInputObject;
+
+// Given a list of shared input objects, returns the starting cost of a transaction that operates on
+// these objects.
+//
+// Starting cost is a proxy for the starting time of the transaction. It is determined by all the input
+// shared objects' last write.
+pub fn compute_tx_start_at_cost(
+    object_execution_cost: &HashMap<ObjectID, u64>,
+    shared_input_objects: &[SharedInputObject],
+) -> u64 {
+    let mut start_cost = 0;
+    shared_input_objects.iter().for_each(|obj| {
+        let object_cost = *object_execution_cost.get(&obj.id).unwrap_or(&0);
+        start_cost = max(start_cost, object_cost);
+    });
+    start_cost
+}
+
+// Given a transaction, returns the deferral key and the congested objects if the transaction should be deferred.
+pub fn should_defer_due_to_object_congestion(
+    object_execution_cost: &HashMap<ObjectID, u64>,
+    cert: &VerifiedExecutableTransaction,
+    max_accumulated_txn_cost_per_object_in_checkpoint: u64,
+    previously_deferred_tx_digests: &HashMap<TransactionDigest, DeferralKey>,
+    commit_round: Round,
+) -> Option<(DeferralKey, Vec<ObjectID>)> {
+    let shared_input_objects: Vec<_> = cert.shared_input_objects().collect();
+    let start_cost = compute_tx_start_at_cost(object_execution_cost, &shared_input_objects);
+    if start_cost + cert.gas_budget() <= max_accumulated_txn_cost_per_object_in_checkpoint {
+        return None;
+    }
+
+    // Finds out the congested objects.
+    //
+    // Note that the congested objects here may be caused by transaction dependency of other congested objects.
+    // Consider in a consensus commit, there are many transactions touching object A, and later in processing the
+    // consensus commit, there is a transaction touching both object A and B. Although there are fewer transactions
+    // touching object B, becase it's starting execution is delayed due to dependency to other transactions on
+    // object A, it may be shown up as congested objects.
+    let mut congested_objects = vec![];
+    for obj in cert.shared_input_objects() {
+        if &start_cost == object_execution_cost.get(&obj.id).unwrap_or(&0) {
```

Comment:
Why does it have to be `==` for this to be congested? For instance, you could have an object whose starting cost is `start_cost - 1`, but if you add gas_budget to it, it will also go over the threshold

Reply:
Very good question! Here is my thought which may not be valid.

When I was thinking about this, I was mainly thinking about execution critical path, and hoping that single hot shared object, or a common set of hot shared objects in a workload are the common case (the case mentioned in the above comment is rare). Therefore, the hot object will always be in the critical path of execution, also means that if it is not on the critical path, it may not actually be hot.

I was thinking for those object in the current consensus commit with starting_cost < txn_start_cost, but starting_cost + gas_budget > limit, if they are actually hot, statistically they should show up as congested object for some transactions (even though not the current one). If they are not shown up in any, they may suffer from the case in the above comment.

Sorry that I was not explain this clearly in the code, because this is kinda related to giving gas price suggestion, and I was mainly thinking about using recent congested object statistics, but TBH I don't have a concrete idea yet. (also the congested objects are not used in this PR yet).

Let me think more about how we want to calculate suggested gas price. If we need to add those start_cost - 1 objects, I'll do it in a separate PR. I also need to add test for it.